nanoPU_sim.py

```diff
@@ -435,33 +420,28 @@ def creditEvent(self, tx_msg_id, new_credit = None, dif_credit = None,
             elif opCode == None:
                 self.log('ERROR: creditEvent was triggered without a opCode value!')
             elif opCode == 'write':
-                if new_credit != None:
-                    if relOp(compVal, cur_credit):
-                        self.log('Changing credit for msg {} from {} to {}'.format(tx_msg_id, cur_credit, new_credit))
-                        self.credit[tx_msg_id] = new_credit
-                else:
-                    self.log("ERROR: creditEvent was triggered with 'write' opCode but no 'new_credit' value is provided!")
+                if relOp(compVal, cur_credit):
+                    self.log('Changing credit for msg {} from {} to {}'.format(tx_msg_id, cur_credit, new_credit))
+                    self.credit[tx_msg_id] = new_credit
             elif opCode == 'add':
-                if dif_credit != None:
-                    if relOp(compVal, cur_credit):
-                        self.credit[tx_msg_id] += dif_credit
-                        self.log('Changing credit for msg {} from {} to {}'.format(tx_msg_id, cur_credit, self.credit[tx_msg_id]))
-                else:
-                    self.log("ERROR: creditEvent was triggered with 'add' opCode but no 'dif_credit' value is provided!")
+                if relOp(compVal, cur_credit):
+                    self.credit[tx_msg_id] += new_credit
+                    self.log('Changing credit for msg {} from {} to {}'.format(tx_msg_id, cur_credit, self.credit[tx_msg_id]))
             elif opCode == 'shift_right':
-                if dif_credit != None:
-                    if relOp(compVal, cur_credit):
-                        self.credit[tx_msg_id] >>= dif_credit
-                        self.log('Changing credit for msg {} from {} to {}'.format(tx_msg_id, cur_credit, self.credit[tx_msg_id]))
-                else:
-                    self.log("ERROR: creditEvent was triggered with 'shift_right' opCode but no 'dif_credit' value is provided!")
+                if relOp(compVal, cur_credit):
+                    self.credit[tx_msg_id] >>= new_credit
+                    self.log('Changing credit for msg {} from {} to {}'.format(tx_msg_id, cur_credit, self.credit[tx_msg_id]))
             else:
                 self.log('ERROR: creditEvent was triggered for unknown opCode: {}'.format(opCode))
 
-            if self.toBtx[tx_msg_id] & (1<<self.credit[tx_msg_id])-1 != 0:
-                self.enq_active_messages_fifo(tx_msg_id)
-        else:
-            self.log('ERROR: creditEvent was triggered for unknown tx_msg_id: {}'.format(tx_msg_id))
+            # NOTE: here's another Read-Modify-Write to update toBtx state, which
+            #   should be combined with the above operation in the actual HW implementation
+            tx_pkts = self.toBtx[tx_msg_id] & (1<<self.credit[tx_msg_id])-1
+            if tx_pkts != 0:
+                # schedule the pkts for transmission
+                self.scheduled_pkts_fifo.put((tx_msg_id, tx_pkts))
```

Comment:
I should admit that this is a clever idea. For the paper we write in the future, we should add the following disclaimer into the paper though:
> When the msg is pushed into the scheduled_pkts_fifo, the toBtx packets are locked for this iteration of the message. By the time the message is dequeued, the message may have additional packets that also have have toBtx bits set. Although those additional packets might have lower sequence number (higher priority for transmission) they are not transmitted until the scheduled_pkts_fifo re-enqueues this message. This may slightly alter the arrival order of packets, but this alteration would be limited to a small number of packets. (highest alteration amount is RTTbytes -1 packets for NDP)

Reply:
Is the situation you are concerned about something like this: a PULL pkts arrives and a new pkt is scheduled for tx (e.g. pkt 6) then immediately after the PULL pkt, a NACK arrives indicating that pkt 4 should be retransmitted. In which case, pkt 6 is transmitted before pkt 4?

Reply:
Yep

Reply:
Yeah this might cause some re-ordering of generated DATA pkts when we compare to the actual NDP implementation. Something to keep in mind.